pkg/logql/log/labels.go

```diff
@@ -1,19 +1,137 @@
-type Labels map[string]string
+import (
+	"sort"
 
-func (l Labels) Has(key string) bool {
-	_, ok := l[key]
-	return ok
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// LabelsBuilder is the same as labels.Builder but tailored for this package.
+type LabelsBuilder struct {
+	base labels.Labels
+	del  []string
+	add  []labels.Label
+
+	err string
+}
+
+// NewLabelsBuilder creates a new labels builder.
+func NewLabelsBuilder() *LabelsBuilder {
+	return &LabelsBuilder{
+		del: make([]string, 0, 5),
+		add: make([]labels.Label, 0, 5),
+	}
+}
+
+// Reset clears all current state for the builder.
+func (b *LabelsBuilder) Reset(base labels.Labels) {
+	b.base = base
+	b.del = b.del[:0]
+	b.add = b.add[:0]
+	b.err = ""
 }
 
-func (l Labels) SetError(err string) {
-	l[ErrorLabel] = err
+// SetErr sets the error label.
+func (b *LabelsBuilder) SetErr(err string) *LabelsBuilder {
+	b.err = err
+	return b
 }
 
-func (l Labels) HasError() bool {
-	_, ok := l[ErrorLabel]
-	return ok
+// GetErr return the current error label value.
+func (b *LabelsBuilder) GetErr() string {
+	return b.err
+}
+
+// HasErr tells if the error label has been set.
+func (b *LabelsBuilder) HasErr() bool {
+	return b.err != ""
+}
+
+// Base returns the base labels unmodified
+func (b *LabelsBuilder) Base() labels.Labels {
+	return b.base
+}
+
+func (b *LabelsBuilder) Get(key string) (string, bool) {
+	for _, a := range b.add {
+		if a.Name == key {
+			return a.Value, true
+		}
+	}
+Outer:
+	for _, l := range b.base {
+		if l.Name == key {
+			for _, d := range b.del {
+				if d == key {
+					continue Outer
+				}
+			}
+			return l.Value, true
+		}
+	}
+	return "", false
+}
+
+// Del deletes the label of the given name.
+func (b *LabelsBuilder) Del(ns ...string) *LabelsBuilder {
+	for _, n := range ns {
+		for i, a := range b.add {
+			if a.Name == n {
+				b.add = append(b.add[:i], b.add[i+1:]...)
+			}
+		}
+		b.del = append(b.del, n)
+	}
+	return b
+}
+
+// Set the name/value pair as a label.
+func (b *LabelsBuilder) Set(n, v string) *LabelsBuilder {
+	for i, a := range b.add {
+		if a.Name == n {
+			b.add[i].Value = v
+			return b
+		}
+	}
+	b.add = append(b.add, labels.Label{Name: n, Value: v})
+
+	return b
+}
+
+// Labels returns the labels from the builder. If no modifications
+// were made, the original labels are returned.
+func (b *LabelsBuilder) Labels() labels.Labels {
+	if len(b.del) == 0 && len(b.add) == 0 {
+		if b.err == "" {
+			return b.base
+		}
+		res := append(b.base, labels.Label{Name: ErrorLabel, Value: b.err})
+		sort.Sort(res)
+		return res
+	}
+
+	// In the general case, labels are removed, modified or moved
+	// rather than added.
+	res := make(labels.Labels, 0, len(b.base))
+Outer:
+	for _, l := range b.base {
+		for _, n := range b.del {
+			if l.Name == n {
+				continue Outer
+			}
+		}
+		for _, la := range b.add {
+			if l.Name == la.Name {
+				continue Outer
+			}
+		}
+		res = append(res, l)
+	}
+	res = append(res, b.add...)
```

Comment:
What about labels that are present in `b.add` and `b.del`, but not in `b.base`? Wouldn't we erroneously add them?

Reply:
if you look at how we add and del, you'll see we keep them clean., they never have clash